File: ios/Modules/Sources/VortexCrypto/impl/Argon2idHasher.swift
import Foundation
import Argon2Swift

/// Argon2id (RFC 9106) via the `Argon2Swift` bridge over libargon2.
///
/// Passwords arrive as `Data` (UTF-8) so the caller can zero the buffer
/// after the call. The hash is derived deterministically from
/// (password, salt, params), so the caller persists [Argon2Params] with
/// the salt to verify later without extra metadata.
public final class Argon2idHasher: PasswordHasher {
    public init() {}

    public func hash(password: Data, salt: Data, params: Argon2Params) throws -> Data {
        let saltObj = Salt(bytes: salt)
        let result = try Argon2Swift.hashPasswordBytes(
            password: password,
            salt: saltObj,
            iterations: Int(params.iterations),
            memory: Int(params.memoryKiB),
            parallelism: Int(params.parallelism),
            length: Int(params.hashLen),
            type: Argon2Type.id,
            version: Argon2Version.V13,
        )
        return result.hashData()
    }

    public func verify(password: Data, salt: Data, expected: Data, params: Argon2Params) throws -> Bool {
        let got = try hash(password: password, salt: salt, params: params)
        return constantTimeEquals(got, expected)
    }

    private func constantTimeEquals(_ a: Data, _ b: Data) -> Bool {
        guard a.count == b.count else { return false }
        var diff: UInt8 = 0
        for i in 0..<a.count { diff |= a[i] ^ b[i] }
        return diff == 0
    }
}

File: android/app/src/main/java/sol/vortexx/android/crypto/impl/Argon2idHasher.kt
package sol.vortexx.android.crypto.impl

import org.bouncycastle.crypto.generators.Argon2BytesGenerator
import org.bouncycastle.crypto.params.Argon2Parameters
import sol.vortexx.android.crypto.api.Argon2Params
import sol.vortexx.android.crypto.api.PasswordHasher
import java.nio.CharBuffer
import java.nio.charset.StandardCharsets
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Argon2id — RFC 9106. Hash is derived deterministically from
 * (password, salt, params) so a stored hash can be re-verified without
 * separately recording iterations/memory (the caller persists [Argon2Params]
 * alongside the salt).
 *
 * Passwords arrive as CharArray and are zeroed after the UTF-8 round-trip.
 */
@Singleton
class Argon2idHasher @Inject constructor() : PasswordHasher {

    override fun hash(password: CharArray, salt: ByteArray, params: Argon2Params): ByteArray {
        val pwBytes = charArrayToUtf8(password)
        try {
            val p = Argon2Parameters.Builder(Argon2Parameters.ARGON2_id)
                .withVersion(Argon2Parameters.ARGON2_VERSION_13)
                .withSalt(salt)
                .withIterations(params.iterations)
                .withMemoryAsKB(params.memoryKb)
                .withParallelism(params.parallelism)
                .build()
            val gen = Argon2BytesGenerator().apply { init(p) }
            val out = ByteArray(params.hashLen)
            gen.generateBytes(pwBytes, out)
            return out
        } finally {
            pwBytes.fill(0)
        }
    }

    override fun verify(
        password: CharArray,
        salt: ByteArray,
        expected: ByteArray,
        params: Argon2Params,
    ): Boolean {
        val got = hash(password, salt, params)
        return constantTimeEquals(got, expected)
    }

    // ── internals ──────────────────────────────────────────────────────

    private fun charArrayToUtf8(chars: CharArray): ByteArray {
        // Don't go through String — that would pin an immutable copy in the
        // intern pool, defeating the point of accepting CharArray.
        val buf = StandardCharsets.UTF_8.encode(CharBuffer.wrap(chars))
        val out = ByteArray(buf.remaining()).also { buf.get(it) }
        return out
    }

    private fun constantTimeEquals(a: ByteArray, b: ByteArray): Boolean {
        if (a.size != b.size) return false
        var diff = 0
        for (i in a.indices) diff = diff or (a[i].toInt() xor b[i].toInt())
        return diff == 0
    }
}
